drivers/i2c/i2c_rtio_default.c

```diff
@@ -12,97 +12,126 @@
-	msgs[0].len = iodev_sqe->sqe.tx.buf_len;
-	msgs[0].flags =
+	msg->buf = (uint8_t *)iodev_sqe->sqe.tx.buf;
+	msg->len = iodev_sqe->sqe.tx.buf_len;
+	msg->flags =
 		((iodev_sqe->sqe.iodev_flags & RTIO_IODEV_I2C_STOP) ? I2C_MSG_STOP : 0) |
 		((iodev_sqe->sqe.iodev_flags & RTIO_IODEV_I2C_RESTART) ? I2C_MSG_RESTART : 0) |
 		((iodev_sqe->sqe.iodev_flags & RTIO_IODEV_I2C_10_BITS) ? I2C_MSG_ADDR_10_BITS : 0) |
 		I2C_MSG_WRITE;
-	*num_msgs = 1;
-	return 0;
 }
 
-static int i2c_iodev_submit_tiny_tx(struct rtio_iodev_sqe *iodev_sqe, struct i2c_msg msgs[2],
-				    uint8_t *num_msgs)
+static inline void i2c_msg_from_tiny_tx(const struct rtio_iodev_sqe *iodev_sqe, struct i2c_msg *msg)
 {
 	__ASSERT_NO_MSG(iodev_sqe->sqe.op == RTIO_OP_TINY_TX);
 
-	msgs[0].buf = (uint8_t *)iodev_sqe->sqe.tiny_tx.buf;
-	msgs[0].len = iodev_sqe->sqe.tiny_tx.buf_len;
-	msgs[0].flags =
+	msg->buf = (uint8_t *)iodev_sqe->sqe.tiny_tx.buf;
+	msg->len = iodev_sqe->sqe.tiny_tx.buf_len;
+	msg->flags =
 		((iodev_sqe->sqe.iodev_flags & RTIO_IODEV_I2C_STOP) ? I2C_MSG_STOP : 0) |
 		((iodev_sqe->sqe.iodev_flags & RTIO_IODEV_I2C_RESTART) ? I2C_MSG_RESTART : 0) |
 		((iodev_sqe->sqe.iodev_flags & RTIO_IODEV_I2C_10_BITS) ? I2C_MSG_ADDR_10_BITS : 0) |
 		I2C_MSG_WRITE;
-	*num_msgs = 1;
-	return 0;
 }
 
-void i2c_iodev_submit_work_handler(struct rtio_iodev_sqe *iodev_sqe)
+void i2c_iodev_submit_work_handler(struct rtio_iodev_sqe *txn_first)
 {
-	const struct i2c_dt_spec *dt_spec = (const struct i2c_dt_spec *)iodev_sqe->sqe.iodev->data;
+	const struct i2c_dt_spec *dt_spec = (const struct i2c_dt_spec *)txn_first->sqe.iodev->data;
 	const struct device *dev = dt_spec->bus;
 
-	LOG_DBG("Sync RTIO work item for: %p", (void *)iodev_sqe);
-
-	struct rtio_iodev_sqe *transaction_current = iodev_sqe;
-	struct i2c_msg msgs[2];
-	uint8_t num_msgs;
+	LOG_DBG("Sync RTIO work item for: %p", (void *)txn_first);
+	uint32_t num_msgs = 0;
 	int rc = 0;
+	struct rtio_iodev_sqe *txn_last = txn_first;
 
+	/* We allocate the i2c_msg's on the stack, to do so
+	 * the count of messages needs to be determined to
+	 * ensure we don't go over the statically sized array.
+	 */
 	do {
-		/* Convert the iodev_sqe back to an i2c_msg */
-		switch (transaction_current->sqe.op) {
+		switch (txn_last->sqe.op) {
+		case RTIO_OP_RX:
+		case RTIO_OP_TX:
+		case RTIO_OP_TINY_TX:
+			num_msgs++;
+			break;
+		default:
+			LOG_ERR("Invalid op code %d for submission %p", txn_last->sqe.op,
+				(void *)&txn_last->sqe);
+			rc = -EIO;
+			break;
+		}
+		txn_last = rtio_txn_next(txn_last);
+	} while (rc == 0 && txn_last != NULL);
+
+	if (rc != 0) {
+		rtio_iodev_sqe_err(txn_first, rc);
+		return;
+	}
+
+	/* Allocate msgs on the stack, MISRA doesn't like VLAs so we need a statically
+	* sized array here. It's pretty unlikely we have more than 4 i2c messages
+	* in a transaction as we typically would only have 2, one to write a
+	* register address, and another to read/write the register into an array
+	*/
+	if (num_msgs > 4) {
```

Comment:
maybe overlay pedantic, but maybe define the 4? `#define RTIO_I2C_MSG_ALLOC 4` and even better, make it a Kconfig?

Reply:
yes, done